### PR TITLE
Use physical properties manager

### PR DESCRIPTION
**Pull request by blaisb**
# Description of the problem

- The Physical Property Manager was developed and well implemented, but it was not used yet in the assemblers and scratch

# Description of the solution

- Add the usage of the physical Property manager in 

- [x] Heat Transfer
- [x] Tracer
- [x] Fluid Dynamics
- [x] VANS

- Add the physical properties model for tracer diffusivity and thermal expansion coefficient

# How Has This Been Tested?

- If all previous tests are working, I think everything is a go
- Added a new test for linear thermal conductivity (gls_navier_stokes_2d/heat_transfer_linear_thermal_conductivity.prm) which tests a steady-state heat problem with a linear thermal conductivity (k= 1+2*T). It converges pretty well. This enabled me to fix some bugs

# Documentation

Documentation has been adapted to take into account the changes.

# Future changes

-  Integrate the locally evaluated density and viscosity for the CFD-DEM models.

### CODE REVIEW

File: include/core/physical_property_model.h
Comment (CaroleAnneDaunais) [inline include/core/physical_property_model.h:44]: Does this mean that the given fields parameter already contains an element for each possible id (i.e. shear_rate, temperature and previous_temperature)? What I mean is, when this function will be called, is the fields parameters something that looks like this:
`{
{shear_rate, []},
{temperature, []},
{previous_temperature, []},
}` ?
If yes, why would it already contain all 3 fields? If not, the at() method will not work if the field is not already existing. The [ ] accessor on the other had will get the proper field if already existing, but will create it if it does not exist. We'll discuss that tomorrow :)

File: include/core/physical_property_model.h
Comment (CaroleAnneDaunais) [inline include/core/physical_property_model.h:48]: The std::vector class has it's own copy constructor. It could look something like:
`std::vector<double> target(fields[id]); `

File: include/core/physical_property_model.h
Comment (CaroleAnneDaunais) [reply include/core/physical_property_model.h:44]: I think I better understand now that the specified id has to be in the fields.

File: source/solvers/heat_transfer_assemblers.cc
Comment (CaroleAnneDaunais) [inline source/solvers/heat_transfer_assemblers.cc:17]: Goodbye simplification to the Navier-Stokes equations

File: source/solvers/heat_transfer_assemblers.cc
Comment (CaroleAnneDaunais) [inline source/solvers/heat_transfer_assemblers.cc:50]: Will this be taken care of in the VOF assembler?

File: source/solvers/heat_transfer_scratch_data.cc
Comment (CaroleAnneDaunais) [inline source/solvers/heat_transfer_scratch_data.cc:200]: Love this

File: include/core/physical_property_model.h
Comment (blaisb) [reply include/core/physical_property_model.h:48]: Yeah but here we don't want to allocate a new vector right, just to copy the data into it.We want to avoid dynamic allocation of vectors when assembling the matrices.

File: include/core/physical_property_model.h
Comment (blaisb) [reply include/core/physical_property_model.h:44]: It's a very good point. The field vector has to have that memory pre-allocated, otherwise it won't work. I faced a challenge here. I wanted to have a flexible structure, but then I ended up not wanting to do dynamic memory allocation in the scratch reinit, because that's the whole point of the scratch. So my best solution was to allocate the memory from the get_go for the scratch, but only fill the fields that we would need.

File: source/solvers/heat_transfer_scratch_data.cc
Comment (CaroleAnneDaunais) [inline source/solvers/heat_transfer_scratch_data.cc:77]: What to these lines do that's different from lines 36 to 39? I think I'm a bit confused

Comment (CaroleAnneDaunais) [reply source/solvers/heat_transfer_scratch_data.cc:77]: Awesome I'll make the changes for Navier-stokes :)

**Comment by blaisb** [comment]
@CaroleAnneDaunais  I think everything is migrated. However, I am not very satisfied with how density is handled right now. It's making me unsure if we should make it as a property.
I propose we review the solver as-is now, as this PR is already getting our of hand. once it's merged, I'll continue the cleanup. I'll start a github project page to keep note of what needs to be cleaned.

**Comment by blaisb** [comment]
@CaroleAnneDaunais  I added defensive assertions in the locations where the non-constant properties are not acceptable yet. Future work will be to extend these things, but I'd rather move onward with this PR now instead of making it crazy large. Right now everything works  and things are, I believe, simpler :).

File: source/solvers/tracer_scratch_data.cc
Comment (jeannejoachim) [inline source/solvers/tracer_scratch_data.cc:72]: Why did you choose to create two variables, rather than a vector containing them? (thus having `diffusivity_model[0]` or something like that)

File: include/core/thermal_expansion_model.h
Comment (jeannejoachim) [inline include/core/thermal_expansion_model.h:58]: why not `inline` here?

File: include/core/thermal_expansion_model.h
Comment (jeannejoachim) [inline include/core/thermal_expansion_model.h:59]: why `/*fields_value*/` is commented? Does this mean that right now nothing is parsed, but it should be the `fields_values` eventually? (so it's a WIP)

File: include/core/thermal_expansion_model.h
Comment (jeannejoachim) [inline include/core/thermal_expansion_model.h:23]: So it's abstract, just as `AuxiliaryPhysics`? If yes, should not the methods be `virtual`?

File: include/solvers/physical_properties_manager.h
Comment (jeannejoachim) [inline include/solvers/physical_properties_manager.h:37]: why is this line to long? It should have been cut by the clang formatting, no?

File: include/solvers/physical_properties_manager.h
Comment (jeannejoachim) [inline include/solvers/physical_properties_manager.h:45]: same as above

**Review by jeannejoachim — APPROVED** [review]
What a PR! I have few questions here and there. Overall I think the code really gained a lot of clarity, well done for this new structure!

File: source/solvers/gd_navier_stokes.cc
Comment (shahabgol) [inline source/solvers/gd_navier_stokes.cc:771]: maybe add a comment to explain what temporary_rheology is doing?

File: source/solvers/gls_navier_stokes.cc
Comment (shahabgol) [inline source/solvers/gls_navier_stokes.cc:729]: the same comment here.

File: source/solvers/heat_transfer_assemblers.cc
Comment (shahabgol) [inline source/solvers/heat_transfer_assemblers.cc:315]:
```suggestion
      const double rho_cp = density[q] * specific_heat[q];
```

File: source/solvers/heat_transfer_assemblers.cc
Comment (shahabgol) [inline source/solvers/heat_transfer_assemblers.cc:494]:
```suggestion
      const double dynamic_viscosity = viscosity[q] * density[q];
```

File: include/solvers/physical_properties_manager.h
Comment (blaisb) [reply include/solvers/physical_properties_manager.h:37]: Good question. I'll manually break it into multiple lines.

File: source/solvers/tracer_scratch_data.cc
Comment (blaisb) [reply source/solvers/tracer_scratch_data.cc:72]: I don't have any good reason for that. I guess that was mostly laziness? I could make a routine for the property_manager that returns the vector of density model straight away and that could be a good way to work. I'll fix that.

File: source/solvers/tracer_scratch_data.cc
Comment (blaisb) [reply source/solvers/tracer_scratch_data.cc:69]: Agreed, defensive programming is good programming. I'll add that.

File: source/fem-dem/vans_assemblers.cc
Comment (shahabgol) [inline source/fem-dem/vans_assemblers.cc:62]: store viscosity[q] as a local variable?

File: source/fem-dem/vans_assemblers.cc
Comment (shahabgol) [inline source/fem-dem/vans_assemblers.cc:237]: same comment.

File: source/fem-dem/vans_assemblers.cc
Comment (shahabgol) [inline source/fem-dem/vans_assemblers.cc:377]: same comment.

File: source/fem-dem/vans_assemblers.cc
Comment (shahabgol) [inline source/fem-dem/vans_assemblers.cc:543]: same comment.

File: include/core/physical_property_model.h
Comment (blaisb) [reply include/core/physical_property_model.h:39]: Inlining getters is generally a good idea. However, if the function has inheritance (e..g class B derives from class A) inlining will not occur because of the need to go through a vtable. So the TLDR would be : If the function is used by the class or the class has no inheritance tree, inline functions in the .h if they are short, otherwise dont :).

File: source/fem-dem/vans_assemblers.cc
Comment (shahabgol) [inline source/fem-dem/vans_assemblers.cc:965]:
```suggestion
                           (viscosity + DBL_MIN);
```

File: include/core/thermal_expansion_model.h
Comment (blaisb) [reply include/core/thermal_expansion_model.h:23]: I don't think the virtual keyword here is necessary. Since the class does not implement the value, vector_value and etc. of the base class, I don't think it could be instantiated (although I might be wrong ) :).

File: source/solvers/navier_stokes_assemblers.cc
Comment (shahabgol) [inline source/solvers/navier_stokes_assemblers.cc:214]: same comment.

File: source/solvers/navier_stokes_assemblers.cc
Comment (shahabgol) [inline source/solvers/navier_stokes_assemblers.cc:515]: same comment.

File: source/fem-dem/vans_assemblers.cc
Comment (blaisb) [reply source/fem-dem/vans_assemblers.cc:965]: The viscosity is always defined positive, why add a DBL_MIN? I don't think the model makes much sense with viscosity = 0 ?

File: include/core/thermal_expansion_model.h
Comment (jeannejoachim) [reply include/core/thermal_expansion_model.h:59]: interesting, I did not know this usage!

File: source/fem-dem/vans_assemblers.cc
Comment (shahabgol) [reply source/fem-dem/vans_assemblers.cc:965]: In the 3d simulation I'm doing to find the pressure difference between inside and outside of a bubble, the viscosity of the phases are equal to zero (so that the NS simulations can be solved analytically). But if you think it's unnecessary, we can ignore it.

File: include/solvers/navier_stokes_scratch_data.h
Comment (blaisb) [reply include/solvers/navier_stokes_scratch_data.h:728]: It is not necessary anymore because the scratch can decide if the hessians are required from the rheologicla model.

File: source/solvers/multiphysics_interface.cc
Comment (blaisb) [reply source/solvers/multiphysics_interface.cc:9]: I managed to revert this back and make it const. It is back to being const now, like it should have been all along.

File: source/solvers/tracer_scratch_data.cc
Comment (blaisb) [reply source/solvers/tracer_scratch_data.cc:69]: Made  it a switch. Cleaner.

File: source/fem-dem/vans_assemblers.cc
Comment (blaisb) [reply source/fem-dem/vans_assemblers.cc:965]: Ohh ok, I did not think about it this way. I'll add this then.

File: source/solvers/gd_navier_stokes.cc
Comment (blaisb) [reply source/solvers/gd_navier_stokes.cc:771]: Agreed :)!. Done

**Comment by blaisb** [comment]
I applied everyone's comment. Thank you for the good reviews :). I'll merge tomorrow once CI has passed.

**Comment by blaisb** [comment]
Documentation has been finalized this morning :).

**Pull request merged**
